File: python_leaoai/python_leaoai/repositories/Implementations/Factory/ProjectGroupFactory.py
# repositories/project_group_factory.py
from typing import List, Optional
from sqlalchemy.orm import Session
from models.factory import ProjectGroup

class ProjectGroupFactory:
    def __init__(self, db: Session):
        self.db = db

    # --- CRUD básico ---
    def get_obj(self, id: int) -> Optional[ProjectGroup]:
        return self.db.query(ProjectGroup).filter(ProjectGroup.id == id).first()

    def save_obj(self, obj: ProjectGroup) -> ProjectGroup:
        self.db.add(obj)
        self.db.commit()
        self.db.refresh(obj)
        return obj

    def update_obj(self, obj: ProjectGroup) -> ProjectGroup:
        self.db.merge(obj)
        self.db.commit()
        self.db.refresh(obj)
        return obj

    def get_all(self) -> List[ProjectGroup]:
        return self.db.query(ProjectGroup).all()

    def get_list(self) -> List[ProjectGroup]:
        return self.db.query(ProjectGroup).filter(ProjectGroup.excluido == False).all()

    def delete_obj(self, id: int) -> bool:
        obj = self.get_obj(id)
        if not obj:
            return False
        try:
            obj.ativo = False
            obj.excluido = True
            self.db.commit()
            return True
        except:
            self.db.rollback()
            return False

    # --- Consultas customizadas ---
    def get_groups_by_project_phase(self, project_phase_id: int) -> List[ProjectGroup]:
        return self.db.query(ProjectGroup)\
            .filter(ProjectGroup.project_phase_id == project_phase_id, ProjectGroup.excluido == False)\
            .all()

    def get_groups_by_project(self, project_id: int) -> List[ProjectGroup]:
        return self.db.query(ProjectGroup)\
            .filter(ProjectGroup.project_id == project_id, ProjectGroup.excluido == False)\
            .all()

File: Framework.Repositories/Implementations/Factory/ProjectGroupFactory.cs
using System.Collections.Generic;
using System.Linq;
using Framework.Data.Models.Factory;

namespace Data.Repositories
{
	public class ProjectGroupFactory
	{
		private readonly ApplicationDbContext _context;

		public ProjectGroupFactory(ApplicationDbContext context)
		{
			_context = context;
		}

		#region Crud
		public ProjectGroup GetObj(int id)
		{
			return _context.ProjectGroups.FirstOrDefault(pg => pg.Id == id);
		}

		public ProjectGroup SaveObj(ProjectGroup obj)
		{
			_context.ProjectGroups.Add(obj);
			_context.SaveChanges();
			return obj;
		}

		public ProjectGroup UpdateObj(ProjectGroup obj)
		{
			_context.ProjectGroups.Update(obj);
			_context.SaveChanges();
			return obj;
		}

		public List<ProjectGroup> GetAll()
		{
			return _context.ProjectGroups.ToList();
		}

		public List<ProjectGroup> GetList()
		{
			return _context.ProjectGroups.Where(pg => !pg.Excluido).ToList();
		}

		public bool DeleteObj(int id)
		{
			var obj = GetObj(id);
			if (obj == null)
				return false;
			try
			{
				obj.Ativo = false;
				obj.Excluido = true;
				_context.ProjectGroups.Update(obj);
				_context.SaveChanges();
				return true;
			}
			catch (Exception)
			{
				return false;
			}
		}
		#endregion

		#region Custom Queries
		public List<ProjectGroup> GetGroupsByProjectPhase(int projectPhaseId)
		{
			return _context.ProjectGroups
						   .Where(pg => pg.ProjectPhaseId == projectPhaseId && !pg.Excluido)
						   .ToList();
		}

		public List<ProjectGroup> GetGroupsByProject(int projectId)
		{
			return _context.ProjectGroups
						   .Where(pg => pg.ProjectId == projectId && !pg.Excluido)
						   .ToList();
		}
		#endregion
	}
}
